pkg/controller/upgradeconfig/cluster_upgrader.go

```diff
@@ -0,0 +1,768 @@
+	err := c.Get(context.TODO(), types.NamespacedName{Name: "version"}, clusterVersion)
+	if err != nil {
+		return false, err
+	}
+	// https://issues.redhat.com/browse/OSD-3442
+	clusterVersion.Spec.Overrides = []configv1.ComponentOverride{}
+	clusterVersion.Spec.DesiredUpdate = &configv1.Update{Version: upgradeConfig.Spec.Desired.Version}
+	err = c.Update(context.TODO(), clusterVersion)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+// TODO This create the maint window for control plane
+func CreateControlPlaneMaintWindow(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return true, nil
+}
+
+// TODO This remove the maint window for control plane
+func RemoveControlPlaneMaintWindow(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return true, nil
+}
+
+// TODO This create the maint window for control plane
+func CreateWorkerMaintWindow(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return true, nil
+}
+
+// This check whether all the master nodes are ready with new config
+func AllMastersUpgraded(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return NodesUpgraded(c, "master", reqLogger)
+
+}
+
+// This check whether all the worker nodes are ready with new config
+func AllWorkersUpgraded(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+
+	return NodesUpgraded(c, "worker", reqLogger)
+
+}
+
+// This will apply extra fixes for different version
+func PostUpgradeFix(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+	fix := postFixes[upgradeConfig.Spec.Desired.Version]
+	if fix != nil {
+		log.Info(fmt.Sprintf("running post fixes for version: %s", upgradeConfig.Spec.Desired.Version))
+		return fix(c, upgradeConfig, reqLogger)
+	}
+
+	return true, nil
+}
+
+// This will remove the extra worker nodes we added before kick off upgrade
+func RemoveExtraScaledNodes(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+	upgradeMachinesets := &machineapi.MachineSetList{}
+
+	err := c.List(context.TODO(), upgradeMachinesets, []client.ListOption{
+		client.InNamespace("openshift-machine-api"),
+		client.MatchingLabels{LABEL_UPGRADE: "true"},
+	}...)
+	if err != nil {
+		reqLogger.Error(err, "failed to get upgrade extra machinesets")
+		return false, err
+	}
+	for _, item := range upgradeMachinesets.Items {
+		err = c.Delete(context.TODO(), &item)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	return true, nil
+}
+
+// This will update the 3rd subscriptions
+func UpdateSubscriptions(c client.Client, upgradeConfig *upgradev1alpha1.UpgradeConfig, reqLogger logr.Logger) (bool, error) {
+	for _, item := range upgradeConfig.Spec.SubscriptionUpdates {
```

Comment:
Does this need to be validated too?

Reply:
I am not sure, if we need verify that, maybe we can do that in https://github.com/openshift/managed-cluster-validating-webhooks/

Reply:
Let's use this https://issues.redhat.com/browse/OSD-3915